ui/vr_webgl.js

```diff
@@ -0,0 +1,611 @@
+      this.gl_.texImage2D(this.gl_.TEXTURE_2D, 0, this.gl_.RGBA,
+          this.gl_.RGBA, this.gl_.UNSIGNED_BYTE, this.video_);
+    }
+
+    // Update matrix
+    shaka.ui.Matrix4x4.multiply(this.viewProjectionMatrix_,
+        this.viewMatrix_, this.identityMatrix_);
+    shaka.ui.Matrix4x4.multiply(this.viewProjectionMatrix_,
+        this.projectionMatrix_, this.viewProjectionMatrix_);
+
+    // Plumbing
+    // Vertices
+    this.gl_.bindBuffer(this.gl_.ARRAY_BUFFER, this.verticesBuffer_);
+    goog.asserts.assert(this.vertexPositionAttribute_ != null,
+        'Should have a texture attribute!');
+    this.gl_.vertexAttribPointer(
+        this.vertexPositionAttribute_, 3, this.gl_.FLOAT, false, 0, 0);
+    this.gl_.enableVertexAttribArray(this.vertexPositionAttribute_);
+
+    // UVs
+    this.gl_.bindBuffer(
+        this.gl_.ARRAY_BUFFER, this.verticesTextureCoordBuffer_);
+    goog.asserts.assert(this.textureCoordAttribute_ != null,
+        'Should have a texture attribute!');
+    this.gl_.vertexAttribPointer(
+        this.textureCoordAttribute_, 2, this.gl_.FLOAT, false, 0, 0);
+    this.gl_.enableVertexAttribArray(this.textureCoordAttribute_);
+
+    this.gl_.bindBuffer(
+        this.gl_.ELEMENT_ARRAY_BUFFER, this.verticesIndexBuffer_);
+
+    this.setMatrixUniforms_();
+
+    this.gl_.uniform1i(
+        this.gl_.getUniformLocation(this.shaderProgram_, 'uSampler'), 0);
+
+    if (this.stereoscopicMode_) {
+      this.gl_.viewport(0, 0, this.canvas_.width / 2, this.canvas_.height);
+    }
+
+    // Draw
+    this.gl_.drawElements(this.gl_.TRIANGLES,
+        this.sphere_.indices.length, this.gl_.UNSIGNED_SHORT, 0);
+
+    if (this.stereoscopicMode_) {
+      this.gl_.viewport(this.canvas_.width / 2, 0,
+          this.canvas_.width / 2, this.canvas_.height);
+      this.gl_.drawElements(this.gl_.TRIANGLES,
+          this.sphere_.indices.length, this.gl_.UNSIGNED_SHORT, 0);
+    }
+  }
+
+  /**
+   * @private
+   */
+  setMatrixUniforms_() {
+    const pUniform =
+        this.gl_.getUniformLocation(this.shaderProgram_, 'u_VPMatrix');
+    this.gl_.uniformMatrix4fv(pUniform, false, this.viewProjectionMatrix_);
+  }
+
+  /**
+   * @private
+   */
+  updateViewPort_() {
+    const scale = 1.8;
+    let currentWidth = this.video_.videoWidth;
+    if (!currentWidth) {
+      currentWidth = this.canvas_.scrollWidth;
+    }
+    let currentHeight = this.video_.videoHeight;
+    if (!currentHeight) {
+      currentHeight = this.canvas_.scrollHeight;
+    }
+
+    if (this.previousCanvasWidth_ !== currentWidth ||
+        this.previousCanvasHeight_ !== currentHeight) {
+      this.canvas_.width = currentWidth;
+      this.canvas_.height = currentHeight;
+
+      this.previousCanvasWidth_ = currentWidth;
+      this.previousCanvasHeight_ = currentHeight;
+
+      const ratio = currentWidth / currentHeight;
+
+      this.projectionMatrix_ = shaka.ui.Matrix4x4.frustum(
+          this.projectionMatrix_, -ratio, ratio, -1, 1, scale, 50.5);
```

Comment:
Also, are 1.8 and 50.5 arbitrary values, or are they required for this projection to work in some way?

Reply:
You found a bug with test values :)